### ISSUE TITLE
Runing pitclipse with Eclipse 2023-12 and Java 21 fails

### ISSUE THREAD

**Issue by NolwennD** [pitest/pitclipse#221]
## Bug description

Java 21 is not supported by the packaged pitest (1.6.8)
```
Exception in thread "main" java.lang.IllegalArgumentException: Unsupported class file major version 65
```

## Expected behavior

Run mutation testing

## How to reproduce
Have a JDK 21 on your computer.

```
git clone git@github.com:NolwennD/UglyMolKKy-RefactoringKata.git

import existing maven project from java folder

run pitest on MolkkyTest
```

## Additional context

Everything is fine with pitest and pitest-maven  in version 1.15.3 with this command:
```
mvn test-compile org.pitest:pitest-maven:mutationCoverage
```

- **Pitclipse version**:
Installed from marketplace (v2.2.1.v20230330-0839)
- **Eclipse IDE version**:
- 2023-12

**Comment by LorenzoBettini** [comment]
@NolwennD thanks for reporting this!
I guess it's time to update the version of pit we're using in pitclipse.

**Comment by LorenzoBettini** [comment]
@echebbi do you have some time to have a look at this or provide some hints?

**Comment by echebbi** [comment]
@LorenzoBettini sure, I'll take a look!

**Comment by LorenzoBettini** [comment]
@echebbi thanks! I tried to port it to the new version of pit, but there was an API breakage and I wasn't able to adapt the runtime code on which I don't have internal knowledge.

**Comment by LorenzoBettini** [comment]
@echebbi , first of all, thanks for the hard work on porting :)

I think we should follow what PIT does, and forget about them for the moment. (especially if these mutators are not recommended). I'd like to keep Pitclipse as simple as possible in that respect.

Please, let me know if I can help somehow.

I saw that you moved the Java version to 21 in GitHub Actions. Maybe that's not required: I seem to understand that the new version of PIT supports Java 21, but not requires that (hopefully). I bet Java 21 will give problems with the old version of Tycho we're still using. In that respect, in a separate PR, I can update our build to use the new version of Tycho (4.x), which requires Java 17 for the build. What do you think? I seem to remember, though, that some unit tests fail with Java 17 due to its new way of dealing with streams and files, but that could be adjusted.

**Comment by LorenzoBettini** [comment]
@echebbi concerning the failure in the CI (I see you have already moved to Tycho 4), you need to use this step to setup a more recent version of Maven required by Tycho 4:

```yaml
    - name: Set up Maven
      uses: stCarolas/setup-maven@v5
      with:
        maven-version: 3.9.6
```

**Comment by echebbi** [comment]
@LorenzoBettini Thanks, I'm not familiar with GitHub Actions so that's helpful!

> I saw that you moved the Java version to 21 in GitHub Actions. Maybe that's not required

I agree, it was just a quick workaround because [the CI was failing due to IProgressMonitor being compiled with Java 17](https://github.com/pitest/pitclipse/actions/runs/8403993120/job/23015098509#step:7:2318). That's surprising because the environment should be set by the target platform. Please let me know if you have any pointer on that issue.

**Comment by LorenzoBettini** [comment]
> I agree, it was just a quick workaround because [the CI was failing due to IProgressMonitor being compiled with Java 17](https://github.com/pitest/pitclipse/actions/runs/8403993120/job/23015098509#step:7:2318). That's surprising because the environment should be set by the target platform. Please let me know if you have any pointer on that issue.

From the log I seem to understand it's actually Tycho itself that somehow takes a version of progress monitor from a recent version of an Eclipse maven artifact: it still hasn't started to resolve the target platform, it happens in the early stage. In any case, we should upgrade the build to Java 17 anyway now :)

**Comment by echebbi** [comment]
@LorenzoBettini fine by me, feel free to update Java and Tycho in a dedicated PR and I'll update my branch once the PR is merged

**Comment by LorenzoBettini** [comment]
@echebbi OK, but did you find a way to fix the unit tests related to files in Java 17?

**Comment by echebbi** [comment]
@LorenzoBettini I'm not sure, are you talking about the failing tests in `ObjectStreamSocketTest`? I figured out that we can remove the call to `spy`, which fixes the tests:

![image](https://github.com/pitest/pitclipse/assets/25926433/2e992fa9-08c9-496f-902b-83f01c874d3e)

The input stream doesn't seem to be used as a spy anyway so I believe that `spy` is a leftover from older tests and can be safely removed.

**Comment by LorenzoBettini** [comment]
@echebbi I remembered a few failing tests, a few months ago, but I don't remember. I'll check with your modification.
